### PR TITLE
network: fix issues arround DHCP DUID-UUID

### LINKED ISSUE #18996: DHCPv6 DUID options are under DHCPv4 section
**systemd version the issue has been seen with**
 > $ systemctl --version
systemd 247 (247.3-1-arch)
+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)

 > `Linux hostname 5.11.0-seblu #1 SMP PREEMPT Tue Feb 16 15:56:00 CET 2021 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
The DHCPv6 DUID related configurations settings in the `[DHCPv6]` section.

Network providers sometime require a custom DUID to obtain a lease. The DHCPv6 client use  `ClientIdentifier`, `DUIDType` and `DUIDRawData` to compute the DUID to send to the DHCPv6 server. This use to be in `[DHCP]` section, but with the split into `[DHCPv4]` and `[DHCPv6]` parameters, the `.network` files should look like this:
```ini
[Network]
DHCP=ipv6

[DHCPv6]
ClientIdentifier=duid-only
DUIDType=link-layer
DUIDRawData=35:36:37:38:39:40:41:42
WithoutRA=solicit
```

**Unexpected behaviour you saw**

The previous `.network` configuration produce *Unknown key name* messages.
```
# SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/systemd-networkd
sd-device-monitor: No permission to stat PID1's netns, unable to determine if we are in host netns: Permission denied
Bus bus-api-network: changing state UNSET → OPENING
sd-bus: starting bus bus-api-network by connecting to /run/dbus/system_bus_socket...
Bus bus-api-network: changing state OPENING → AUTHENTICATING
Registering bus object implementation for path=/org/freedesktop/LogControl1 iface=org.freedesktop.LogControl1
timestamp of '/etc/systemd/network' changed
/etc/systemd/network/ilovelennart.network:20: Unknown key name 'ClientIdentifier' in section 'DHCPv6', ignoring.
/etc/systemd/network/ilovelennart.network:21: Unknown key name 'DUIDType' in section 'DHCPv6', ignoring.
/etc/systemd/network/ilovelennart.network:23: Unknown key name 'DUIDRawData' in section 'DHCPv6', ignoring.
```

And the working configuration use `[DHCPv4]` in order to configure DHCPv6 clients.

```ini
[Network]
DHCP=ipv6

[DHCPv4]
ClientIdentifier=duid-only
DUIDType=link-layer
DUIDRawData=35:36:37:38:39:40:41:42

[DHCPv6]
WithoutRA=solicit
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #18996.

### CODE REVIEW

**Comment by bluca** [comment]
```
--- command ---
04:50:03 /home/runner/work/systemd/systemd/tools/check-directives.sh /home/runner/work/systemd/systemd /home/runner/work/systemd/systemd/build
--- stdout ---
67,68d66
< DUIDRawData=
< DUIDType=
154d151
< IAID=
Looks like test/fuzz/fuzz-network-parser/directives.network hasn't been updated
```

**Comment by yuwata** [comment]
Oh, I forgot to update the directive for the fuzzer. PTAL.

**Comment by eyJhb** [comment]
Thank you ! I can test this when I get home later today if you want me to :)

**Comment by yuwata** [comment]
@eyJhb Thank you!

**Comment by yuwata** [comment]
Hmm. test-logind fails on Ubuntu CIs... I have no idea why. But it seems caused by this PR...

**Comment by eyJhb** [comment]
I am having a hard time testing this, since I am running NixOS and our current patches for systemd will not apply clean for this systemd branch. I will continue to try some more, to see if I can get it to work :)

**Comment by yuwata** [comment]
I do not know why, but now test-logind passes.

File: src/network/networkd-dhcp-common.c [hunk 1]
```diff
@@ -101,6 +101,42 @@ static int duid_set_uuid(DUID *duid, sd_id128_t uuid) {
...
+        int r;
+
+        assert(link);
+
+        if (!IN_SET(link->state, LINK_STATE_CONFIGURING, LINK_STATE_CONFIGURED))
+                return 0;
+
+        log_link_info(link, "%s", __func__);
+
+        r = dhcp4_configure(link, false);
```
Comment (yuwata) [inline src/network/networkd-dhcp-common.c]: `if (!link->dhcp_client) {`

File: src/network/networkd-dhcp-common.c [hunk 2]
```diff
@@ -101,6 +101,42 @@ static int duid_set_uuid(DUID *duid, sd_id128_t uuid) {
...
+        if (r < 0)
+                return r;
+
+        r = dhcp6_configure(link, false);
```
Comment (yuwata) [inline src/network/networkd-dhcp-common.c]: `if (!link->dhcp6_client) {`

**Review by yuwata — COMMENTED** [review]
I will fix the followings later.

File: src/network/networkd-dhcp-common.c [hunk 3]
```diff
@@ -64,15 +64,26 @@ void network_adjust_dhcp(Network *network) {
 }
 
 static struct DUID fallback_duid = { .type = DUID_TYPE_EN };
-DUID* link_get_duid(Link *link) {
-        if (link->network->duid.type != _DUID_TYPE_INVALID)
-                return &link->network->duid;
-        else if (link->hw_addr.length == 0 && IN_SET(link->manager->duid.type, DUID_TYPE_LLT, DUID_TYPE_LL))
+
+DUID* link_get_duid(Link *link, int family) {
```
Comment (poettering) [inline src/network/networkd-dhcp-common.c]: can we make the return struct `const`here btw? if so, we cann make fallback_duid const too, which sounds apropriate given it's actually constant, no?

Comment (yuwata) [reply src/network/networkd-dhcp-common.c]: Done.

**Comment by eyJhb** [comment]
Sorry for bothering, is there any timeline for when this will be in a tagged release?

**Comment by yuwata** [comment]
@eyJhb v249-rc1.

**Pull request merged**